### ISSUE TITLE
FasterRCNN example inference on CPU

### ISSUE THREAD

**Issue by susueah** [tensorpack/tensorpack#1024]
I trained FasterRCNN model on my dataset and now wanted to do inference on CPU

However there is an assert at the beginning of inference that suggests that this is not supported
`assert tf.test.is_gpu_available()` 

If I remove it and  run inference on CPU I get

`InvalidArgumentError (see above for traceback): Default MaxPoolingOp only supports NHWC on device type CPU
         [[node pool0/MaxPool (defined at /home/ubuntu/miniconda3/lib/python3.6/site-packages/tensorpack/models/pool.py:35)  = MaxPool[T=DT_FLOAT, data_format="NCHW", ksize=[1, 1, 3, 3], padding="VALID", strides=[1, 1, 2, 2], _device="/job:localhost/replica:0/task:0/device:CPU:0"](Pad_1)]]`

This probably means that FasterRCNN uses `data_format` that is not supported on CPU.

Is it possible, with a slight change to a model code, to make it work on CPU, or did I miss something and CPU is actually supported?

**Comment by ppwwyyxx** [comment]
CPU is not supported and a small change is not enough.

**Comment by Remember2018** [comment]
Hi @ppwwyyxx , thanks a lot for your great tensorpack! I have  a question: do you mean the Faster-RCNN model trained with tensorpack cannot be depolyed in CPU environment? Thanks.

**Comment by ppwwyyxx** [comment]
"A small change is not enough" implies that with a large change this is possible.

It can be deployed if you write the model in a way that tensorflow supports on CPU.

**Comment by Remember2018** [comment]
@ppwwyyxx thanks a lot for your reply! Could you please give some advice on how to do this? Do we need to replace all the `maxpool(NCHW)` with `maxpool(NHWC)`? Thanks.

**Comment by Remember2018** [comment]
Yeah, thanks a lot. I checked the `FasterRCNN/model_fpn.py`, it seems you have shown the solution from https://github.com/tensorpack/tensorpack/blob/b097e7d5acefa4eec13873f257035ac0c3db595b/examples/FasterRCNN/model_fpn.py#L40-L45

And I think we may modify all the backbone graph and the FasterRCNN graph to replace the MaxPooling(channel_first) to MaxPooling(channel_last).

So I think maybe the easist way is to implement the Maxpooling(NCHW) in CPU mode. I'm not sure why this is not implemented in TensorFlow.

**Comment by Remember2018** [comment]
Hi @ppwwyyxx , it seems the `tf.nn.max_pool` support the NCHW format. Please check this issue https://github.com/tensorflow/tensorflow/issues/15364#issuecomment-394766712 

Does this mean the tf.layers APIs used in tensorpack have some inconsistency with the tf.nn APIs?

**Comment by ppwwyyxx** [comment]
>  it seems the tf.nn.max_pool support the NCHW format

No. Not on CPUs.

**Comment by ppwwyyxx** [comment]
The code does not work for me or @susueah on CPU. 
Maybe it works in some version.

If it NCHW works for you already, sounds good and you may have less work to do then.

**Comment by Remember2018** [comment]
Thank you @ppwwyyxx . I used the tf-1.10.1. I'm still figuring out why the saved model trained with tensorpack cannot be deployed in tensorflow-serving CPU environment. 

I checked the tf.layers.MaxPooling2D API in 

https://github.com/tensorflow/tensorflow/blob/4dcfddc5d12018a5a0fdca652b9221ed95e9eb23/tensorflow/python/keras/layers/pooling.py#L304-L308

it seems the function used the `tf.nn.max_pool`, if the code snippts above can be run in CPU mode, why not the tf.layers.MaxPooling2D? I probably missed something.

**Comment by ppwwyyxx** [comment]
It's more likely that your tensorflow is built with MKL which includes support for NCHW layout, however ours are the default build without MKL.

In any case such issues do not seem to be related to tensorpack and I don't have any other comments on your questions.

**Comment by Remember2018** [comment]
Thanks a lot! @ppwwyyxx , I just used the pip install from official website.

**Comment by ppwwyyxx** [comment]
The lines are documentation. Not code.

**Comment by Remember2018** [comment]
Although this might be unrelated to TensorPack, please allow me to post what I have tried here. I tried the following methods, but both failed:

(1) compile the tensorflow and the tensorflow_model_server with Intel-MKL support, then using the tensorflow-serving-api to serve the FasterRCNN model trained with TensorPack. But the following error occurs:

    raise _Rendezvous(state, None, None, deadline)
    grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.ABORTED, Operation received an exception:Status: 3, message: could not create a dilated convolution forward descriptor, in file external/org_tensorflow/tensorflow/core/kernels/mkl_conv_ops.cc:1125
         [[{{node conv0/Conv2D}}]])>

(2) replace all the `MaxPooling2D` and `tf.nn.avg_pool` with NCHW in the FasterRCNN model with the corresponding NHWC versions (plus the extra two `tf.transpose` operations).  But the following error occurs:

    raise _Rendezvous(state, None, None, deadline)
    grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.UNIMPLEMENTED, Generic conv implementation only supports NHWC tensor format for now.
         [[{{node conv0/Conv2D}} = Conv2D[T=DT_FLOAT, _output_shapes=[[1,64,?,?]], data_format="NCHW", dilations=[1, 1, 1, 1], padding="VALID", strides=[1, 1, 2, 2], use_cudnn_on_gpu=true, _device="/job:localhost/replica:0/task:0/device:CPU:0"](Pad, conv0/W/read)]])>

So the final way is, retrain the ImageNet-Pretrained ResNet models with NHWC data format, and then modify all the NCHW operations to the NHWC version? 

Is it possible to re-create the graph with all operations in NHWC format, and then change all the variables in NCHW format to the NHWC format?

**Comment by ppwwyyxx** [comment]
Your error log is about convolution, not max pooling. Only replacing maxpooling is certainly not enough.
You never need to retrain the model because both data_format uses the weights in the same format.

**Comment by Remember2018** [comment]
@ppwwyyxx thanks to your kind help, I have modified the FasterRCNN with NHWC format, and now the CPU inference of FasterRCNN is OK. 
But the speed seems slow. For testing one resized image with short edge size of 800 and max size of 1333， the time of ResNet-50-FPN is about 2.5 seconds.

**Comment by roadcode** [comment]
@Remember2018 how do yo modify the FasterRCNN with NHWC format?

**Comment by Remember2018** [comment]
@roadcode In fact, @ppwwyyxx has posted one solution in https://github.com/tensorpack/tensorpack/tree/fpn_cpu_inference/examples/FasterRCNN branch.

**Comment by ppwwyyxx** [comment]
Without a GPU, inference can also run in NCHW format if tensorflow is built with MKL support.

**Comment by Remember2018** [comment]
Hi, I'm using ResNet50 backbone and FasterRCNN. The inference time for one image (resized to 800x1333) in one GTX 1080Ti is about 200ms. Is this reasonable? But I remembered that others can achieve to 70ms. Could anybody give some advice on how to speed up the inference? Thanks.

**Comment by ppwwyyxx** [comment]
It seems to be a reasonable speed.
Tensorpack is a library for model training and does not provide optimization of inference speed.

**Comment by fuzzyBatman** [comment]
@Remember2018 
Can you please make a summary of the steps required to train and run inference on the CPU? It would be of great help.
